README.md
# Mock.js
<!-- 模擬請求 & 模擬數據 -->
[![Build Status](https://travis-ci.org/nuysoft/Mock.svg?branch=refactoring)](https://travis-ci.org/nuysoft/Mock)

<!-- [![Coverage Status](https://coveralls.io/repos/nuysoft/Mock/badge.png?branch=refactoring)](https://coveralls.io/r/nuysoft/Mock?branch=refactoring)
[![NPM version](https://badge.fury.io/js/mockjs.svg)](http://badge.fury.io/js/mockjs)
[![Bower version](https://badge.fury.io/bo/mockjs.svg)](http://badge.fury.io/bo/mockjs)
[![Dependency Status](https://gemnasium.com/nuysoft/Mock.svg)](https://gemnasium.com/nuysoft/Mock)
[![spm package](http://spmjs.io/badge/mockjs)](http://spmjs.io/package/mockjs) -->

Mock.js is a simulation data generator to help the front-end to develop and prototype separate from the back-end progress and reduce some monotony particularly while writing automated tests.

The official site: <http://mockjs.com>

## Features

* Change the original resource file to Traditional Chinese
* Generate simulated data according to the data template
* Provide request/response mocking for ajax requests

## Ｏriginal Source
https://github.com/nuysoft/Mock

## License
Mock.js is available under the terms of the [MIT License](./LICENSE).

CHANGELOG.md
# Changelog

## 2015.12.03 V0.2.0-alpha2

1. 改用 webpack 打包

### 2014.3.20 V0.2.0-alpha1

1. 增加網站

### 2014.12.23 V0.2.0 重構代碼

1. 改用 gulp 打包
2. 改用 mocha 重寫測試用例
3. 改用 requirejs 重構代碼

### 2014.6.24 V0.2.0 重構代碼

1. 支持 UMD，包括：
    * 未打包前的代碼
    * 打包後的代碼
2. random CLI
    * --help 增加方法和參數說明
3. 重構文檔站 @蘿素
    * 增加《入門》
    * 單列《文檔》
4. 測試用例
    * 重寫測試用例
    * 同時支持 nodeunit 和 qunit
    * 同時支持 jQuery、KISSY、Zepto
    * 同時支持 KMD、AMD、CMD
5. 覆寫 XHR @行列 @霍庸
6. 廢棄的功能
    * Mock.mockjax()
    * Mock.tpl()
    * Mock.xtpl()
7. Random.dateImage() 支持 node-canvas
8. Mock.valid(tpl, data)
9. Mock.toJOSNSchema()
10. Mock.mock(regexp) 
11. 完善地支持 node，代碼中的：
    * window
    * document
    * XHRHttpRequest
12. 支持相對路徑

### 2014.6.23 V0.1.5

1. [!] 修覆 #28 #29，因為 jQuery 每個版本在 Ajax 實現上有些差異，導致在攔截 Ajax 請求時出現了兼容性問題（例如，方法 `xhr.onload()` 訪問不到）。本次[測試](http://jsfiddle.net/8y8Fz/)並通過的 jQuery 版本有：

    * jQuery 2.1.0
    * jQuery 2.0.2
    * jQuery 1.11.0
    * jQuery 1.10.1
    * jQuery 1.9.1
    * jQuery 1.8.3
    * jQuery 1.7.2
    * jQuery 1.6.4

非常抱歉，這個問題一直困擾著 Mock.js 用戶，在後面的版本中，會通過攔截 XMLHttpRequest 的方法“一勞永逸”地解決攔截 Ajax 的兼容和適配問題。

### 2014.6.18 V0.1.4

1. [!] 修覆 #14 0.1.1版本試了好像jq1.10可以，1.11下$.ajax攔截沒反應
2. [!] 修覆 #22 異步加載js文件的時候發現問題
3. [!] 修覆 #23 Mock.mockjax 導致 $.getScript 不執行回調
4. [!] 修覆 #24 Window Firefox 30.0 引用 占位符 拋錯
5. [!] 修覆 #25 改變了非函數屬性的順序，查找起來不方便
6. [!] 修覆 #26 生成規則 支持 負數 number|-100-+100
7. [!] 修覆 #27 數據模板編輯器 格式化（Tidy） 時會丟掉 函數屬性
8. [+] 數據模板編輯器 增加了 編輯區 和 生成結果區 的同步滾動
9. [!] test/nodeuinit > test/nodeunit

### 2014.5.26 V0.1.3

1. [!] 修覆 #21

### 2014.5.26 V0.1.2

1. [!] 重構 Mock.mockjax()
2. [!] 更新 package.json/devDependencies
3. [+] 增加 懶懶交流會 PPT

### 2014.5.9 V0.1.2
1. [+] 支持 [`Mock.mock(rurl, rtype, template)`](http://mockjs.com/#mock)
2. [+] 支持 [`'name|min-max': {}`、`'name|count': {}`](http://mockjs.com/#語法規範)
3. [+] 支持 [`'name': function(){}`](http://mockjs.com/#語法規範)
4. [+] 新增占位符 [@NOW](http://mockjs.com/#now)
5. [+] 更新了 [語法規範](http://mockjs.com/#語法規範)

### 2013.9.6
1. 增加占位符 @DATAIMAGE
2. 解析占位符時**完全**忽略大小寫

### 2013.9.3
1. 文檔增加用法示例：Sea.js (CMD)、RequireJS (AMD)
2. 增加對 CMD 規範的支持
3. 生成 SourceMap 文件 `dist/mock-min.map`

### 2013.8.21
1. 100% 基於客戶端模板生成模擬數據，支持 KISSY XTemplate。
1. 調整文件結構。

### 2013.8.11
1. 80% 基於客戶端模板生成模擬數據。
1. 完善針對 KISSY XTemplate 的測試用例 [test/mock4tpl-xtpl-node.js](test/mock4tpl-xtpl-node.js)。
1. [Mock4Tpl](src/tpl/mock4tpl.js) 支持 Partials。
1. Mock 支持轉義 @。
1. 更新 README.md，增加對 Mock4Tpl 的說明。
1. 完善 [demo](demo/)。
1. 減少 Mock、Mock4Tpl 暴漏的 API。

### 2013.8.7
1. 75% 基於客戶端模板生成模擬數據。
1. 完善測試用例 [test/mock4tpl-node.js](test/mock4tpl-node.js)。
1. 重構文件和目錄結構，把代碼模塊化。
1. 參考 Handlebars.js，引入 Jison 生成模板解析器。

#### 2013.8.2
1. 60% 基於客戶端模板生成模擬數據。
1. 增加測試用例 [test/mock4tpl-node.js](test/mock4tpl-node.js)，參考自 <http://handlebarsjs.com/>。

#### 2013.7.31
1. 50% 基於客戶端模板生成模擬數據。

#### 2013.7.18
1. 增加占位符 @COLOR。
1. 完善對占位符的解析，過濾掉 `#%&()?/.`。
1. 對“支持的占位符”分組。

#### 2013.7.12
1. Mock.mock(rurl, template) 的參數 rurl 可以是字符串或正則。
1. 把產生隨機元數據的接口封裝到 Mock.Random 中。
1. 增加對日期的格式化。
1. 增加占位符 @IMG、@PARAGRAPH、@SENTENCE、@WORD、@FIRST、@LAST、@NAME、@DOMAIN、@EMAIL、@IP、@ID。
1. 支持嵌套的占位符，例如 `@IMG(@AD_SIZE)`。
1. 支持把普通屬性當作占位符使用，例如 `@IMG(@size)`。

src/mock/handler.js
/* 
    ## Handler

    處理數據模板。
    
    * Handler.gen( template, name?, context? )

        入口方法。

    * Data Template Definition, DTD
        
        處理數據模板定義。

        * Handler.array( options )
        * Handler.object( options )
        * Handler.number( options )
        * Handler.boolean( options )
        * Handler.string( options )
        * Handler.function( options )
        * Handler.regexp( options )
        
        處理路徑（相對和絕對）。

        * Handler.getValueByKeyPath( key, options )

    * Data Placeholder Definition, DPD

        處理數據占位符定義

        * Handler.placeholder( placeholder, context, templateContext, options )

*/

var Constant = require('./constant')
var Util = require('./util')
var Parser = require('./parser')
var Random = require('./random/')
var RE = require('./regexp')

var Handler = {
    extend: Util.extend
}

/*
    template        屬性值（即數據模板）
    name            屬性名
    context         數據上下文，生成後的數據
    templateContext 模板上下文，

    Handle.gen(template, name, options)
    context
        currentContext, templateCurrentContext, 
        path, templatePath
        root, templateRoot
*/
Handler.gen = function (template, name, context) {
    /* jshint -W041 */
    name = name == undefined ? '' : (name + '')

    context = context || {}
    context = {
        // 當前訪問路徑，只有屬性名，不包括生成規則
        path: context.path || [Constant.GUID],
        templatePath: context.templatePath || [Constant.GUID++],
        // 最終屬性值的上下文
        currentContext: context.currentContext,
        // 屬性值模板的上下文
        templateCurrentContext: context.templateCurrentContext || template,
        // 最終值的根
        root: context.root || context.currentContext,
        // 模板的根
        templateRoot: context.templateRoot || context.templateCurrentContext || template
    }
    // console.log('path:', context.path.join('.'), template)

    var rule = Parser.parse(name)
    var type = Util.type(template)
    var data

    if (Handler[type]) {
        data = Handler[type]({
            // 屬性值類型
            type: type,
            // 屬性值模板
            template: template,
            // 屬性名 + 生成規則
            name: name,
            // 屬性名
            parsedName: name ? name.replace(Constant.RE_KEY, '$1') : name,

            // 解析後的生成規則
            rule: rule,
            // 相關上下文
            context: context
        })

        if (!context.root) context.root = data
        return data
    }

    return template
}

Handler.extend({
    array: function (options) {
        var result = [],
            i, ii;

        // 'name|1': []
        // 'name|count': []
        // 'name|min-max': []
        if (options.template.length === 0) return result

        // 'arr': [{ 'email': '@EMAIL' }, { 'email': '@EMAIL' }]
        if (!options.rule.parameters) {
            for (i = 0; i < options.template.length; i++) {
                options.context.path.push(i)
                options.context.templatePath.push(i)
                result.push(
                    Handler.gen(options.template[i], i, {
                        path: options.context.path,
                        templatePath: options.context.templatePath,
                        currentContext: result,
                        templateCurrentContext: options.template,
                        root: options.context.root || result,
                        templateRoot: options.context.templateRoot || options.template
                    })
                )
                options.context.path.pop()
                options.context.templatePath.pop()
            }
        } else {
            // 'method|1': ['GET', 'POST', 'HEAD', 'DELETE']
            if (options.rule.min === 1 && options.rule.max === undefined) {
                // fix #17
                options.context.path.push(options.name)
                options.context.templatePath.push(options.name)
                result = Random.pick(
                    Handler.gen(options.template, undefined, {
                        path: options.context.path,
                        templatePath: options.context.templatePath,
                        currentContext: result,
                        templateCurrentContext: options.template,
                        root: options.context.root || result,
                        templateRoot: options.context.templateRoot || options.template
                    })
                )
                options.context.path.pop()
                options.context.templatePath.pop()
            } else {
                // 'data|+1': [{}, {}]
                if (options.rule.parameters[2]) {
                    options.template.__order_index = options.template.__order_index || 0

                    options.context.path.push(options.name)
                    options.context.templatePath.push(options.name)
                    result = Handler.gen(options.template, undefined, {
                        path: options.context.path,
                        templatePath: options.context.templatePath,
                        currentContext: result,
                        templateCurrentContext: options.template,
                        root: options.context.root || result,
                        templateRoot: options.context.templateRoot || options.template
                    })[
                        options.template.__order_index % options.template.length
                    ]

                    options.template.__order_index += +options.rule.parameters[2]

                    options.context.path.pop()
                    options.context.templatePath.pop()

                } else {
                    // 'data|1-10': [{}]
                    for (i = 0; i < options.rule.count; i++) {
                        // 'data|1-10': [{}, {}]
                        for (ii = 0; ii < options.template.length; ii++) {
                            options.context.path.push(result.length)
                            options.context.templatePath.push(ii)
                            result.push(
                                Handler.gen(options.template[ii], result.length, {
                                    path: options.context.path,
                                    templatePath: options.context.templatePath,
                                    currentContext: result,
                                    templateCurrentContext: options.template,
                                    root: options.context.root || result,
                                    templateRoot: options.context.templateRoot || options.template
                                })
                            )
                            options.context.path.pop()
                            options.context.templatePath.pop()
                        }
                    }
                }
            }
        }
        return result
    },
    object: function (options) {
        var result = {},
            keys, fnKeys, key, parsedKey, inc, i;

        // 'obj|min-max': {}
        /* jshint -W041 */
        if (options.rule.min != undefined) {
            keys = Util.keys(options.template)
            keys = Random.shuffle(keys)
            keys = keys.slice(0, options.rule.count)
            for (i = 0; i < keys.length; i++) {
                key = keys[i]
                parsedKey = key.replace(Constant.RE_KEY, '$1')
                options.context.path.push(parsedKey)
                options.context.templatePath.push(key)
                result[parsedKey] = Handler.gen(options.template[key], key, {
                    path: options.context.path,
                    templatePath: options.context.templatePath,
                    currentContext: result,
                    templateCurrentContext: options.template,
                    root: options.context.root || result,
                    templateRoot: options.context.templateRoot || options.template
                })
                options.context.path.pop()
                options.context.templatePath.pop()
            }

        } else {
            // 'obj': {}
            keys = []
            fnKeys = [] // #25 改變了非函數屬性的順序，查找起來不方便
            for (key in options.template) {
                (typeof options.template[key] === 'function' ? fnKeys : keys).push(key)
            }
            keys = keys.concat(fnKeys)

            /*
                會改變非函數屬性的順序
                keys = Util.keys(options.template)
                keys.sort(function(a, b) {
                    var afn = typeof options.template[a] === 'function'
                    var bfn = typeof options.template[b] === 'function'
                    if (afn === bfn) return 0
                    if (afn && !bfn) return 1
                    if (!afn && bfn) return -1
                })
            */

            for (i = 0; i < keys.length; i++) {
                key = keys[i]
                parsedKey = key.replace(Constant.RE_KEY, '$1')
                options.context.path.push(parsedKey)
                options.context.templatePath.push(key)
                result[parsedKey] = Handler.gen(options.template[key], key, {
                    path: options.context.path,
                    templatePath: options.context.templatePath,
                    currentContext: result,
                    templateCurrentContext: options.template,
                    root: options.context.root || result,
                    templateRoot: options.context.templateRoot || options.template
                })
                options.context.path.pop()
                options.context.templatePath.pop()
                // 'id|+1': 1
                inc = key.match(Constant.RE_KEY)
                if (inc && inc[2] && Util.type(options.template[key]) === 'number') {
                    options.template[key] += parseInt(inc[2], 10)
                }
            }
        }
        return result
    },
    number: function (options) {
        var result, parts;
        if (options.rule.decimal) { // float
            options.template += ''
            parts = options.template.split('.')
            // 'float1|.1-10': 10,
            // 'float2|1-100.1-10': 1,
            // 'float3|999.1-10': 1,
            // 'float4|.3-10': 123.123,
            parts[0] = options.rule.range ? options.rule.count : parts[0]
            parts[1] = (parts[1] || '').slice(0, options.rule.dcount)
            while (parts[1].length < options.rule.dcount) {
                parts[1] += (
                    // 最後一位不能為 0：如果最後一位為 0，會被 JS 引擎忽略掉。
                    (parts[1].length < options.rule.dcount - 1) ? Random.character('number') : Random.character('123456789')
                )
            }
            result = parseFloat(parts.join('.'), 10)
        } else { // integer
            // 'grade1|1-100': 1,
            result = options.rule.range && !options.rule.parameters[2] ? options.rule.count : options.template
        }
        return result
    },
    boolean: function (options) {
        var result;
        // 'prop|multiple': false, 當前值是相反值的概率倍數
        // 'prop|probability-probability': false, 當前值與相反值的概率
        result = options.rule.parameters ? Random.bool(options.rule.min, options.rule.max, options.template) : options.template
        return result
    },
    string: function (options) {
        var result = '',
            i, placeholders, ph, phed;
        if (options.template.length) {

            //  'foo': '★',
            /* jshint -W041 */
            if (options.rule.count == undefined) {
                result += options.template
            }

            // 'star|1-5': '★',
            for (i = 0; i < options.rule.count; i++) {
                result += options.template
            }
            // 'email|1-10': '@EMAIL, ',
            placeholders = result.match(Constant.RE_PLACEHOLDER) || [] // A-Z_0-9 > \w_
            for (i = 0; i < placeholders.length; i++) {
                ph = placeholders[i]

                // 遇到轉義斜杠，不需要解析占位符
                if (/^\\/.test(ph)) {
                    placeholders.splice(i--, 1)
                    continue
                }

                phed = Handler.placeholder(ph, options.context.currentContext, options.context.templateCurrentContext, options)

                // 只有一個占位符，並且沒有其他字符
                if (placeholders.length === 1 && ph === result && typeof phed !== typeof result) { // 
                    result = phed
                    break

                    if (Util.isNumeric(phed)) {
                        result = parseFloat(phed, 10)
                        break
                    }
                    if (/^(true|false)$/.test(phed)) {
                        result = phed === 'true' ? true :
                            phed === 'false' ? false :
                                phed // 已經是布爾值
                        break
                    }
                }
                result = result.replace(ph, phed)
            }

        } else {
            // 'ASCII|1-10': '',
            // 'ASCII': '',
            result = options.rule.range ? Random.string(options.rule.count) : options.template
        }
        return result
    },
    'function': function (options) {
        // ( context, options )
        return options.template.call(options.context.currentContext, options)
    },
    'regexp': function (options) {
        var source = ''

        // 'name': /regexp/,
        /* jshint -W041 */
        if (options.rule.count == undefined) {
            source += options.template.source // regexp.source
        }

        // 'name|1-5': /regexp/,
        for (var i = 0; i < options.rule.count; i++) {
            source += options.template.source
        }

        return RE.Handler.gen(
            RE.Parser.parse(
                source
            )
        )
    }
})

Handler.extend({
    _all: function () {
        var re = {};
        for (var key in Random) re[key.toLowerCase()] = key
        return re
    },
    // 處理占位符，轉換為最終值
    placeholder: function (placeholder, obj, templateContext, options) {
        // console.log(options.context.path)
        // 1 key, 2 params
        Constant.RE_PLACEHOLDER.exec('')
        var parts = Constant.RE_PLACEHOLDER.exec(placeholder),
            key = parts && parts[1],
            lkey = key && key.toLowerCase(),
            okey = this._all()[lkey],
            params = parts && parts[2] || ''
        var pathParts = this.splitPathToArray(key)

        // 解析占位符的參數
        try {
            // 1. 嘗試保持參數的類型
            /*
                #24 [Window Firefox 30.0 引用 占位符 拋錯](https://github.com/nuysoft/Mock/issues/24)
                [BX9056: 各瀏覽器下 window.eval 方法的執行上下文存在差異](http://www.w3help.org/zh-cn/causes/BX9056)
                應該屬於 Window Firefox 30.0 的 BUG
            */
            /* jshint -W061 */
            params = eval('(function(){ return [].splice.call(arguments, 0 ) })(' + params + ')')
        } catch (error) {
            // 2. 如果失敗，只能解析為字符串
            // console.error(error)
            // if (error instanceof ReferenceError) params = parts[2].split(/,\s*/);
            // else throw error
            params = parts[2].split(/,\s*/)
        }

        // 占位符優先引用數據模板中的屬性
        if (obj && (key in obj)) return obj[key]

        // @index @key
        // if (Constant.RE_INDEX.test(key)) return +options.name
        // if (Constant.RE_KEY.test(key)) return options.name

        // 絕對路徑 or 相對路徑
        if (
            key.charAt(0) === '/' ||
            pathParts.length > 1
        ) return this.getValueByKeyPath(key, options)

        // 遞歸引用數據模板中的屬性
        if (templateContext &&
            (typeof templateContext === 'object') &&
            (key in templateContext) &&
            (placeholder !== templateContext[key]) // fix #15 避免自己依賴自己
        ) {
            // 先計算被引用的屬性值
            templateContext[key] = Handler.gen(templateContext[key], key, {
                currentContext: obj,
                templateCurrentContext: templateContext
            })
            return templateContext[key]
        }

        // 如果未找到，則原樣返回
        if (!(key in Random) && !(lkey in Random) && !(okey in Random)) return placeholder

        // 遞歸解析參數中的占位符
        for (var i = 0; i < params.length; i++) {
            Constant.RE_PLACEHOLDER.exec('')
            if (Constant.RE_PLACEHOLDER.test(params[i])) {
                params[i] = Handler.placeholder(params[i], obj, templateContext, options)
            }
        }

        var handle = Random[key] || Random[lkey] || Random[okey]
        switch (Util.type(handle)) {
            case 'array':
                // 自動從數組中取一個，例如 @areas
                return Random.pick(handle)
            case 'function':
                // 執行占位符方法（大多數情況）
                handle.options = options
                var re = handle.apply(Random, params)
                if (re === undefined) re = '' // 因為是在字符串中，所以默認為空字符串。
                delete handle.options
                return re
        }
    },
    getValueByKeyPath: function (key, options) {
        var originalKey = key
        var keyPathParts = this.splitPathToArray(key)
        var absolutePathParts = []

        // 絕對路徑
        if (key.charAt(0) === '/') {
            absolutePathParts = [options.context.path[0]].concat(
                this.normalizePath(keyPathParts)
            )
        } else {
            // 相對路徑
            if (keyPathParts.length > 1) {
                absolutePathParts = options.context.path.slice(0)
                absolutePathParts.pop()
                absolutePathParts = this.normalizePath(
                    absolutePathParts.concat(keyPathParts)
                )

            }
        }

        key = keyPathParts[keyPathParts.length - 1]
        var currentContext = options.context.root
        var templateCurrentContext = options.context.templateRoot
        for (var i = 1; i < absolutePathParts.length - 1; i++) {
            currentContext = currentContext[absolutePathParts[i]]
            templateCurrentContext = templateCurrentContext[absolutePathParts[i]]
        }
        // 引用的值已經計算好
        if (currentContext && (key in currentContext)) return currentContext[key]

        // 尚未計算，遞歸引用數據模板中的屬性
        if (templateCurrentContext &&
            (typeof templateCurrentContext === 'object') &&
            (key in templateCurrentContext) &&
            (originalKey !== templateCurrentContext[key]) // fix #15 避免自己依賴自己
        ) {
            // 先計算被引用的屬性值
            templateCurrentContext[key] = Handler.gen(templateCurrentContext[key], key, {
                currentContext: currentContext,
                templateCurrentContext: templateCurrentContext
            })
            return templateCurrentContext[key]
        }
    },
    // https://github.com/kissyteam/kissy/blob/master/src/path/src/path.js
    normalizePath: function (pathParts) {
        var newPathParts = []
        for (var i = 0; i < pathParts.length; i++) {
            switch (pathParts[i]) {
                case '..':
                    newPathParts.pop()
                    break
                case '.':
                    break
                default:
                    newPathParts.push(pathParts[i])
            }
        }
        return newPathParts
    },
    splitPathToArray: function (path) {
        var parts = path.split(/\/+/);
        if (!parts[parts.length - 1]) parts = parts.slice(0, -1)
        if (!parts[0]) parts = parts.slice(1)
        return parts;
    }
})

module.exports = Handler